libs/libcommon/include/common/avx2_mem_utils.h

```diff
@@ -150,42 +155,64 @@ FLATTEN_INLINE_PURE static inline bool check_block32x4_eq(const char * a, const
 
 FLATTEN_INLINE_PURE static inline int cmp_block32x4(const char * a, const char * b)
 {
-    if (check_block32x4_eq(a, b))
+    if (likely(check_block32x4_eq(a, b)))
         return 0;
-    for (size_t i = 0; i < AVX2_UNROLL_NUM - 1; ++i)
+    for (size_t i = 0; i < 3; ++i)
     {
-        if (auto ret = cmp_block32(a + i * BLOCK32_SIZE, (b + i * BLOCK32_SIZE)); ret)
+        if (auto ret = cmp_block32(a + i * BLOCK32_SIZE, (b + i * BLOCK32_SIZE)); unlikely(ret))
             return ret;
     }
-    return cmp_block32(a + (AVX2_UNROLL_NUM - 1) * BLOCK32_SIZE, (b + (AVX2_UNROLL_NUM - 1) * BLOCK32_SIZE));
-}
-FLATTEN_INLINE_PURE static inline uint32_t swap_u32(uint32_t val)
-{
-    return __builtin_bswap32(val);
-}
-FLATTEN_INLINE_PURE static inline uint64_t swap_u64(uint64_t val)
-{
-    return __builtin_bswap64(val);
-}
-
-[[maybe_unused]] FLATTEN_INLINE_PURE static inline uint32_t read_u32_swap(const void * data)
-{
-    return swap_u32(read<uint32_t>(data));
-}
-
-[[maybe_unused]] FLATTEN_INLINE_PURE static inline uint64_t read_u64_swap(const void * data)
-{
-    return swap_u64(read<uint64_t>(data));
+    return cmp_block32<true>(a + 3 * BLOCK32_SIZE, (b + 3 * BLOCK32_SIZE));
 }
 
 // ref: https://github.com/lattera/glibc/blob/master/sysdeps/x86_64/multiarch/memcmp-avx2-movbe.S
-FLATTEN_INLINE_PURE static inline int avx2_mem_cmp(const char * p1, const char * p2, size_t n)
+FLATTEN_INLINE_PURE static inline int avx2_mem_cmp(const char * p1, const char * p2, size_t n) noexcept
 {
     constexpr size_t loop_block32x4_size = AVX2_UNROLL_NUM * BLOCK32_SIZE;
 
     // n <= 32
     if (likely(n <= BLOCK32_SIZE))
     {
+#ifdef M
+        static_assert(false, "`M` is defined");
+#else
+#define M(x)                             \
+    case (x):                            \
+    {                                    \
+        return std::memcmp(p1, p2, (x)); \
+    }
+#endif
+        switch (n)
+        {
+            M(0);
+            M(1);
+            M(2);
+            M(3);
+            M(4);
+            M(5);
+            M(6);
+            M(7);
+            M(8);
+            M(9);
+            M(10);
+            M(11);
+            M(12);
+            M(13);
+            M(14);
+            M(15);
+            M(16);
+        default:
+        {
+            // 17~32
+            if (auto ret = cmp_block16(p1, p2); ret)
+                return ret;
+            return cmp_block16(p1 + n - BLOCK16_SIZE, p2 + n - BLOCK16_SIZE);
+        }
+        }
+#undef M
+
+// an optional way to check small str
+#if defined(AVX2_MEM_CMP_NORMAL_IF_ELSE)
         if (unlikely(n < 2))
```

Comment:
I'm a little confused: In my understanding if `n <= BLOCK32_SIZE`, it always be handled in the `switch` statement, then when will these code be used?

Reply:
Those code is used for performance test. Now, we choose the `switch case` way for small str, maybe it can be optimized later for other scenarios.